src/core/cluster_analysis/ClusterStructure.cpp

```diff
@@ -0,0 +1,205 @@
+      if (partners!=1) {
+        j+=1+partners;
+        continue;
+      }
+      // We are only here if bond has one partner
+      add_pair(p,partCfg()[p.bl.e[j+1]]);
+      j+=2; // Type id + one prtner
+    }
+  }
+}
+merge_clusters();
+}
+
+
+
+void ClusterStructure::add_pair(const Particle& p1, const Particle& p2) {
+// * check, if there's a neighbor
+ //   * No: Then go on to the next particle
+ // * Yes: Then if
+ //   * One of them belongs to a cluster, give the other one the same cluster
+ //     id.
+ //   * None of them belongs to a cluster: Give them both a new cluster id
+ //   * Both belong to different clusters: Mark the clusters as identical 
+ //   * so that they can be put together later
+  if (! m_pair_criterion) {
+    runtimeErrorMsg() << "No cluster criterion defined"; 
+    return;
+  }
+  // If the two particles are neighbors...
+  if (m_pair_criterion->decide(p1,p2)) {
+     
+     if // None belongs to a cluster
+     ((!part_of_cluster(p1)) && (!part_of_cluster(p2)))
+    {  
+      // Both particles belong to the same, new cluster
+      const int cid=get_next_free_cluster_id();
+
+      // assign the cluster_ids 
+      cluster_id[p1.p.identity]=cid;
+      cluster_id[p2.p.identity]=cid;
+    }
+    else if // p2 belongs to a cluster but p1 doesn't
+      (part_of_cluster(p2) && !part_of_cluster(p1))
+    {
+     // Give p1 the same cluster id as p2
+     cluster_id[p1.p.identity]=find_id_for(cluster_id.at(p2.p.identity));
+    }
+    else if // i belongs to a cluster but j doesn't
+      (part_of_cluster(p1) && !part_of_cluster(p2))
+    {
+     // give p2 the cluster id from p1
+     cluster_id[p2.p.identity]=find_id_for(cluster_id.at(p1.p.identity));
+    }
+    else if // Both belong to different clusters
+      (part_of_cluster(p1) && part_of_cluster(p2) &&
+       cluster_id.at(p1.p.identity)!=cluster_id.at(p2.p.identity))
+    {
+     // Clusters of p1 and p2 are one and the same. Add an identity to the list
+     // The higher number must be inserted as first value of tjhe pair
+     // because the substituions later have to be done in descending order
+     const int cid1=find_id_for(cluster_id.at(p1.p.identity));
+     const int cid2=find_id_for(cluster_id.at(p2.p.identity));
+     if (cid1>cid2)
+     {
+       m_cluster_identities[cid1] =cid2;
+     }
+     else if (cid1<cid2) 
+     {
+       m_cluster_identities[cid2]=cid1;
+     }
+     // else do nothing. The clusters are already noted for merging.
+     // Connected clusters will be merged later
+    }
+    // The case for both particles being in the same cluster does not need to be
+    // treated.
+  }
+}
+
+void ClusterStructure::merge_clusters() {
+  // Relabel particles according to the cluster identities map
+  // Also create empty cluster objects for the final cluster id
+  
+  // Collect needed changes in a separate map, as doing the changes on the fly
+  // would screw up the iterators
+  std::map<int,int> to_be_changed;
```

Comment:
If I understand this correctly you only need lookup by id, and no order. Then you can use std::unordered_map, which has O(N) insert and lookup (std::map is ordered and has O(N*log(N)). I *think* this is also true for clusters.

Reply:
Since the indexing property of `std::map` is used nowhere, it might be better to use
```cpp
std::vector< std::tuple<int,int> > to_be_changed;
```
Then insertion is done via
```cpp
to_be_changed.emplace_back(std::make_tuple(a,b));
```
and iteration is done via
```cpp
for (auto const& it : to_be_changed)
{
    std::get<0>(it);
    std::get<1>(it);
}
```
For more info on why to use `std::vector`, watch [Taming the performance Beast - Klaus Iglberger - Meeting C++ 2015](https://www.youtube.com/watch?v=LrVi9LHP8Bk).

Reply:
@hmenke The ids can have gaps...